lib/paraactions/paraactions.ts

```diff
@@ -0,0 +1,417 @@
+function splitArgs(source: string): string[] {
+  // Use shared top-level splitter for comma-separated arguments
+  // Note: we intentionally do not trim here; parseArgsList is responsible
+  //      for trimming and skipping empty/whitespace-only entries.
+  return splitTopLevel(source, (ch) => ch === ',');
+}
+
+
+// ---------- VALUE PARSING ----------
+
+function parseValue(source: string): ParaValue {
+  const trimmed = source.trim();
+
+  if (!trimmed) {
+    return { kind: 'identifier', value: '' };
+  }
+
+  const firstChar = trimmed[0];
+
+  // String literal
+  if (firstChar === "'" || firstChar === '"') {
+    return parseStringValue(trimmed);
+  }
+
+  // Boolean
+  if (trimmed === 'true') {
+    return { kind: 'boolean', value: true };
+  }
+  if (trimmed === 'false') {
+    return { kind: 'boolean', value: false };
+  }
+
+  // null
+  if (trimmed === 'null') {
+    return { kind: 'null', value: null };
+  }
+
+  // Number
+  const num = Number(trimmed);
+  if (!Number.isNaN(num)) {
+    return { kind: 'number', value: num };
+  }
+
+  // Fallback: identifier (e.g., series names)
+  return { kind: 'identifier', value: trimmed };
+}
+
+function parseStringValue(source: string): ParaValue {
+  const quote = source[0];
+  if (quote !== "'" && quote !== '"') {
+    throw new Error('parseStringValue on non-string: ' + source);
+  }
+
+  if (source.length < 2 || source[source.length - 1] !== quote) {
+    throw new Error('Unterminated string literal: ' + source);
+  }
+
+  const inner = source.slice(1, -1);
+  return {
+    kind: 'string',
+    value: inner,
+  };
+}
+
+// ---------- EXECUTOR (WITH CTX, ERRORS BUBBLE TO HOST) ----------
+
+// execute actions by calling methods directly on the initial context (e.g., parachart.api) ===
+function paraValuesToJsArgs(args: ParaValue[]): unknown[] {
+  return args.map(arg => arg.value);
+}
+
+/**
+ * Execute a single ParaAction (single or chained) by treating each segment
+ * as a method call on the current context object.
+ *
+ * - `initialContext` is typically a root API object such as `parachart.api`.
+ * - Each segment's method name is looked up on the current context.
+ * - If a method returns a non-undefined value, it becomes the context for the next segment.
+ * - If a method is not found on the current context, this function throws.
+ * - This function does NOT catch errors; they bubble to the host.
+ */
+export function executeParaAction<Ctx>(
+  action: ParaAction,
+  initialContext: Ctx
+): Ctx {
+  let ctx: any = initialContext;
+  // console.warn('executeParaAction', ctx)
+  for (const segment of action.segments) {
+    // console.log('ctx', ctx)
+    const method = ctx?.[segment.methodName];
+    if (typeof method !== 'function') {
+      throw new Error(
+        `Unknown action method on context: ${segment.methodName}`
+      );
+    }
+
+    const jsArgs = paraValuesToJsArgs(segment.args);
+    const result = method.apply(ctx, jsArgs);
+    if (typeof result !== 'undefined') {
```

Comment:
This check shouldn't be here. If we had something like `api.getSeries(foo).getPoint(0)`, where 'foo' doesn't exist as a series, then if `getSeries(foo)` is `undefined` (and not `null`), `getPoint(0)` will have the context `api`, giving us a very confusing error message 'Unknown action method on context: getPoint'. If we pass `undefined`s through (and add the error I suggest above) we would get the better error 'Method getPoint called on undefined\null object getSeries(foo)'. Of course, we could avoid this issue by only returning `null` from failed actions, or always throwing, but then this check would be redundant